src/platform/Linux/bluez/ChipDeviceScanner.cpp

```diff
@@ -158,10 +154,10 @@ void ChipDeviceScanner::TimerExpiredCallback(chip::System::Layer * layer, void *
 
 CHIP_ERROR ChipDeviceScanner::StopScan()
 {
-    VerifyOrReturnError(mIsScanning, CHIP_NO_ERROR);
-    VerifyOrReturnError(!mIsStopping, CHIP_NO_ERROR);
+    std::lock_guard<std::mutex> lock(mScannerStoppingLock);
+
+    VerifyOrReturnError(mCurrentState == ChipDeviceScannerState::SCANNER_SCANNING, CHIP_NO_ERROR);
```

Comment:
This change was due to my suggestion :), which was to get rid of the lock completely and make sure that `StopScan` is taking CHIP thread lock (this can be verified by adding `assertChipStackLockedByCurrentThread`). I just thought that most Matter structures are used from a single thread so if there is an exception to that this can be handled externally.
But you maintain this platform, so if you see a reason to use an extra lock, feel free to disregard my suggestion.

Reply:
Yes, you are right, the best way to handle that would be to use the internal CHIP lock to synchronize start/stop calls. But to be honest I'm not sure about the use case that @tianfeng-yang is dealing with. The simplest solution would be to move all the logic in `StartScan()` and `StopScan()` in the functions called with `GLibMatterContextInvokeSync`, since this call serializes everything on matter glib thread.

Reply:
Uh, yeah, I hit a dead end in my thinking.

> The simplest solution would be to move all the logic in `StartScan()` and `StopScan()` in the functions called with `GLibMatterContextInvokeSync`, since this call serializes everything on matter glib thread.

Reply:
Because `StopScan()` may be called in different threads, causing a race condition. As long as `StopScan()` can only be called in the ChipStack thread, the race condition can be solved. Currently `StopScan()` will not cancel the scan timer. If it is called in the ChipStack thread, we can cancel the timer in advance instead of waiting for it to time out.